ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/SnapshotTrieLogManager.java

```diff
@@ -61,38 +63,45 @@ public void addCachedLayer(
         "adding snapshot world state for block {}, state root hash {}",
         blockHeader::toLogString,
         worldStateRootHash::toHexString);
-    worldStateArchive
-        .getMutableSnapshot(worldStateRootHash)
-        .map(BonsaiSnapshotWorldState.class::cast)
-        .ifPresent(
-            snapshot ->
-                cachedWorldStatesByHash.put(
-                    blockHeader.getHash(),
-                    new CachedSnapshotWorldState(snapshot, trieLog, blockHeader.getNumber())));
+    cachedWorldStatesByHash.put(
+        blockHeader.getHash(),
+        new CachedSnapshotWorldState(
+            () ->
+                worldStateArchive
+                    .getMutableSnapshot(blockHeader.getHash())
+                    .map(BonsaiSnapshotWorldState.class::cast)
```

Comment:
I don't understand why this cast is necessary, are there snapshots in the forest implementation?
Maybe the supplier can return optional instead of orElse(null)?